crates/optimism/primitives/src/lib.rs

```diff
@@ -9,3 +9,172 @@
 
 pub mod bedrock;
 pub mod op_tx_type;
+
+use alloy_primitives::{bytes, Bytes, TxKind, Uint, B256};
+
+use alloy_consensus::{SignableTransaction, TxLegacy};
+use alloy_rlp::{Decodable, Encodable, Result};
+use op_tx_type::OpTxType;
+use reth_primitives::revm_primitives::{AccessList, SignedAuthorization};
+use reth_primitives_traits::{InMemorySize, TransactionExt};
+
+use op_alloy_consensus::{OpTxEnvelope, OpTypedTransaction};
+
+#[cfg(feature = "arbitrary")]
+use arbitrary::Arbitrary;
+
+#[cfg_attr(feature = "arbitrary", derive(Arbitrary))]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+/// Optimistic transaction type.
+pub struct OpTransaction(OpTypedTransaction);
+
+impl Default for OpTransaction {
+    fn default() -> Self {
+        Self(OpTypedTransaction::Legacy(TxLegacy::default()))
+    }
+}
+
+impl Encodable for OpTransaction {
+    fn encode(&self, out: &mut dyn bytes::BufMut) {
+        match &self.0 {
+            OpTypedTransaction::Legacy(tx) => tx.encode(out),
+            OpTypedTransaction::Eip2930(tx) => tx.encode(out),
+            OpTypedTransaction::Eip1559(tx) => tx.encode(out),
+            OpTypedTransaction::Eip7702(tx) => tx.encode(out),
+            OpTypedTransaction::Deposit(tx) => tx.encode(out),
+        }
+    }
+}
+
+impl Decodable for OpTransaction {
+    fn decode(data: &mut &[u8]) -> Result<Self> {
+        let tx_envelope = OpTxEnvelope::decode(data)?;
+        Ok(Self(tx_envelope.into()))
+    }
+}
+
+impl reth_codecs::Compact for OpTransaction {
+    fn to_compact<B>(&self, out: &mut B) -> usize
+    where
+        B: bytes::BufMut + AsMut<[u8]>,
+    {
+        match &self.0 {
+            OpTypedTransaction::Legacy(tx) => tx.to_compact(out),
+            OpTypedTransaction::Eip2930(tx) => tx.to_compact(out),
+            OpTypedTransaction::Eip1559(tx) => tx.to_compact(out),
+            OpTypedTransaction::Eip7702(tx) => tx.to_compact(out),
+            OpTypedTransaction::Deposit(tx) => tx.to_compact(out),
+        }
+    }
+
+    fn from_compact(_buf: &[u8], _len: usize) -> (Self, &[u8]) {
+        todo!()
+    }
+}
+
+#[cfg(any(feature = "test-utils", feature = "arbitrary"))]
+impl MaybeArbitrary for OpTransaction {}
+
+#[cfg(any(feature = "test-utils", feature = "arbitrary"))]
+impl MaybeArbitrary for OpTransaction {}
```

Comment:
should be possible to remove this, since there is a blanket impl of `MaybeArbitrary` for all types that impl `Arbitrary`
https://github.com/paradigmxyz/reth/blob/68a6ada460c31c74f25d3b5418b8bc878edfe082/crates/primitives-traits/src/lib.rs#L91-L94
so should be enough just to impl `Arbitrary` (which you have done with a derive macro above)